src/helper.cc

```diff
@@ -573,6 +569,16 @@ Helper::Client::submit(const char * const buf, HLPCB * const callback, void *dat
     debugs(84, DBG_DATA, Raw("buf", buf, strlen(buf)));
 }
 
+void
+Helper::Client::callBack(Xaction *r)
+{
+    const auto callback = r->request.callback;
+    r->request.callback = nullptr;
+    void *cbdata = nullptr;
+    if (cbdataReferenceValidDone(r->request.data, &cbdata))
+        callback(cbdata, r->reply);
+}
```

Comment:
This method assumes `r` is not nil. We probably cannot easily delete `r` here because that would complicate already non-trivial `delete r` conditions in some callers. Thus, we should change this `r` parameter from a pointer to a reference.